sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs

```diff
@@ -466,11 +466,11 @@ private Block BuildTryStatement(TryStatementSyntax tryStatement, Block currentBl
             var hasFinally = tryStatement.Finally?.Block != null;
             if (hasFinally)
             {
-                // Wire exit in case we have a return inside the try/catch block
-                catchSuccessor = BuildBlock(tryStatement.Finally.Block,
-                    CreateBranchBlock(tryStatement.Finally, new[] { catchSuccessor, ExitTarget.Peek() }));
+                // Create a finally block for the happy path where no exceptions are thrown
+                catchSuccessor = BuildBlock(tryStatement.Finally.Block, CreateBlock(catchSuccessor));
 
-                ExitTarget.Push(catchSuccessor);
+                // Wire another finally block to the exit target stack in case we have a return inside the try/catch block
+                ExitTarget.Push(BuildBlock(tryStatement.Finally.Block, CreateBlock(ExitTarget.Peek())));

```

Comment:
Create a separate CFG blocks for finally blocks that are called after catch and after return. This way we are correctly constructing the code paths.